From: Rob Evers <revers@redhat.com>
Date: Mon, 6 Dec 2010 22:42:59 -0500
Subject: [scsi] bfa: fix crash reading driver sysfs statistics
Message-id: <20101206224259.30349.45746.sendpatchset@localhost.localdomain>
Patchwork-id: 29930
O-Subject: [RHEL5.6 PATCH] bfa: driver sysfs crash fix
Bugzilla: 659880
CVE: CVE-2010-4343
RH-Acked-by: Tomas Henzl <thenzl@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=659880

fix crash when access sysfs fc host statistics

Description of problem:
This is the same issue as BZ# 594882 for RHEL 6. The system crashes when read
fc statistics files under /sys/class/fc_host/host#/statistics

The same fix has been accepted by linux upstream and applied to bfa driver in
RHEL 6.0.

Patch provided by Brocade.

I confirmed the fix by inducing the panic without the patch, and then trying
the same after the patch was applied and no panic was seen.

https://brewweb.devel.redhat.com/taskinfo?taskID=2943639
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/bfa/bfa_core.c b/drivers/scsi/bfa/bfa_core.c
index 02488c3..8745ce1 100644
--- a/drivers/scsi/bfa/bfa_core.c
+++ b/drivers/scsi/bfa/bfa_core.c
@@ -88,11 +88,32 @@ bfa_cfg_get_meminfo(struct bfa_iocfc_cfg_s *cfg, struct bfa_meminfo_s *meminfo)
 	for (i = 0; hal_mods[i]; i++)
 		hal_mods[i]->meminfo(cfg, &km_len, &dm_len);
 
+	dm_len += bfa_port_meminfo();
 
 	meminfo->meminfo[BFA_MEM_TYPE_KVA - 1].mem_len = km_len;
 	meminfo->meminfo[BFA_MEM_TYPE_DMA - 1].mem_len = dm_len;
 }
 
+static void
+bfa_com_port_attach(struct bfa_s *bfa, struct bfa_meminfo_s *mi)
+{
+	struct bfa_port_s       *port = &bfa->modules.port;
+	uint32_t                dm_len;
+	uint8_t                 *dm_kva;
+	uint64_t                dm_pa;
+
+	dm_len = bfa_port_meminfo();
+	dm_kva = bfa_meminfo_dma_virt(mi);
+	dm_pa  = bfa_meminfo_dma_phys(mi);
+
+	memset(port, 0, sizeof(struct bfa_port_s));
+	bfa_port_attach(port, &bfa->ioc, bfa, bfa->trcmod, bfa->logm);
+	bfa_port_mem_claim(port, dm_kva, dm_pa);
+
+	bfa_meminfo_dma_virt(mi) = dm_kva + dm_len;
+	bfa_meminfo_dma_phys(mi) = dm_pa + dm_len;
+}
+
 /**
  * Use this function to do attach the driver instance with the BFA
  * library. This function will not trigger any HW initialization
@@ -144,6 +165,7 @@ bfa_attach(struct bfa_s *bfa, void *bfad, struct bfa_iocfc_cfg_s *cfg,
 	for (i = 0; hal_mods[i]; i++)
 		hal_mods[i]->attach(bfa, bfad, cfg, meminfo, pcidev);
 
+	bfa_com_port_attach(bfa, meminfo);
 }
 
 /**
